modules/dnsmasq_dhcp/autodetection.go

```diff
@@ -46,18 +60,128 @@ func (d *DnsmasqDHCP) findDHCPRanges() ([]string, error) {
 	return ranges, nil
 }
 
-func findConfigurations(confDir string) ([]string, error) {
-	fis, err := ioutil.ReadDir(confDir)
+// findConfigs recursively finds and reads configuration files respecting
+// conf-file and conf-dir directives.
+// findConfigs is tolerant to IO errors and finds as maximum config
+// files as possible, therefore if an error occurrs during scanning process,
+// it will be just logged with warning severity
+func (d *DnsmasqDHCP) findConfigs(confPath string) []string {
+	config, err := os.Open(confPath)
+	if err != nil {
+		d.Warningf("error during configuration file %q reading: %v", confPath, err)
+		return nil
+	}
+
+	defer config.Close()
+
+	var (
+		includeFiles []string
+		includeDirs  []configDir
+	)
+
+	scanner := bufio.NewScanner(config)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if path, ok := getConfValue(line, "conf-file"); ok {
+			includeFiles = append(includeFiles, path)
+			continue
+		}
+
+		if path, ok := getConfValue(line, "conf-dir"); ok {
+			args := strings.Split(path, ",")
+
+			dir := configDir{
+				path: args[0],
+			}
+
+			for _, arg := range args[1:] {
+				arg := strings.TrimSpace(arg)
+				// dnsmasq treats suffixes with asterisk as "to include" and without
+				// asterisk as "to exclude"
+				if strings.HasPrefix(arg, "*") {
+					dir.includeSuffix = append(dir.includeSuffix, arg[1:])
+				} else {
+					dir.excludeSuffix = append(dir.excludeSuffix, arg)
+				}
+			}
+
+			includeDirs = append(includeDirs, dir)
+
+			continue
+		}
+	}
+
+	for _, dir := range includeDirs {
+		dirFiles, err := dir.find()
+		if err != nil {
+			d.Warningf("error during configuration dir %q scanning: %v", dir.path, err)
+		}
+
+		includeFiles = append(includeFiles, dirFiles...)
+	}
+
+	for _, file := range includeFiles {
+		files := d.findConfigs(file)
+
+		includeFiles = append(includeFiles, files...)
+	}
+
+	return includeFiles
+}
+
+func getConfValue(line, prefix string) (value string, ok bool) {
+	if !strings.HasPrefix(line, prefix) {
+		return "", false
+	}
+
+	value = strings.TrimPrefix(line, prefix)
+
+	value = strings.TrimSpace(value)
+
+	if !strings.HasPrefix(value, "=") {
+		// got some unexpected line, has prefix like conf-file but there is no
+		// assign sign
+		return "", false
+	}
+
+	value = strings.TrimPrefix(value, "=")
+
+	value = strings.TrimSpace(value)
+
+	return value, true
+}
+
+func (dir configDir) find() ([]string, error) {
+	if len(dir.includeSuffix) == 0 && len(dir.excludeSuffix) == 0 {
+		dir.includeSuffix = []string{".conf"}
+	}
+
+	fis, err := ioutil.ReadDir(dir.path)
 	if err != nil {
 		return nil, err
 	}
 
 	var configs []string
+
+FILES:
 	for _, fi := range fis {
-		if !fi.Mode().IsRegular() || !strings.HasSuffix(fi.Name(), ".conf") {
+		if !fi.Mode().IsRegular() {
 			continue
 		}
-		configs = append(configs, path.Join(confDir, fi.Name()))
+
+		for _, suffix := range dir.excludeSuffix {
+			if strings.HasSuffix(fi.Name(), suffix) {
+				continue FILES
+			}
```

Comment:
// # Include all the files in a directory except those ending in .bak
// #conf-dir=/etc/dnsmasq.d,.bak

This is a little bit more complex then exclude .bak, include .conf

Reply:
and i think `findConfigFiles` is a better name, because function finds config files. Otherwise it is not clear what it finds.

Reply:
i suggest to test how it works when both `exclude` and `include` are set, just to be sure

Reply:
check the last commit